collector/src/main/java/com/bitmovin/analytics/retryBackend/RetryBackend.kt

```diff
@@ -0,0 +1,89 @@
+    fun getNextScheduledTime() = retryQueue.getNextScheduleTime()
+
+    override fun send(eventData: EventData) {
+        scheduleSample(RetrySample(eventData, 0, Date(), 0))
+    }
+
+    override fun sendAd(eventData: AdEventData) {
+        scheduleSample(RetrySample(eventData, 0, Date(), 0))
+    }
+
+    private fun scheduleSample(retrySample: RetrySample<Any>) {
+        val callback = object : OnFailureCallback {
+            override fun onFailure(e: Exception, cancel: () -> Unit) {
+                if (e is SocketTimeoutException || e is ConnectException || e is StreamResetException || e is UnknownHostException) {
+                    cancel()
+                    retryQueue.addSample(retrySample)
+                    processQueuedSamples()
+                }
+            }
+        }
+
+        if (retrySample.eventData is EventData) {
+            Log.d(TAG, "sending sample ${retrySample.eventData?.sequenceNumber} retry ${retrySample.retry}")
+            retrySample.eventData.retryCount = retrySample.retry
+            this.next.send(retrySample.eventData, callback)
+        } else if (retrySample.eventData is AdEventData) {
+            Log.d(TAG, "sending ad sample ${retrySample.eventData?.adId} retry ${retrySample.retry}")
+            retrySample.eventData.retryCount = retrySample.retry
+            this.next.sendAd(retrySample.eventData, callback)
+        }
+    }
+
+    @Synchronized
+    fun processQueuedSamples() {
+        try {
+            val nextScheduledTime = getNextScheduledTime()
+            if (nextScheduledTime != null || retryDateToken != nextScheduledTime) {
+                if (retryDateToken != null) {
+                    scheduleSampleHandler.removeCallbacks(sendNextSampleRunnable, retryDateToken)
+                }
+                retryDateToken = nextScheduledTime
+                val delay = maxOf((nextScheduledTime?.time ?: 0) - Date().time, 0) // to prevent negative delay
+                scheduleSampleHandler.postAtTime(sendNextSampleRunnable, retryDateToken, SystemClock.uptimeMillis() + delay)
+            }
+        } catch (e: Exception) {
+            Log.e(TAG, "processQueuedSamples() threw an unexpected exception: ${e.message}", e)
+        }
+    }
+
+    // https://developer.android.com/reference/java/lang/Runnable
+    private val sendNextSampleRunnable = Runnable {
+        try {
+            val retrySample = retryQueue.getNextSampleOrNull()
+            if (retrySample != null) {
+                scheduleSample(retrySample)
+            }
+            retryDateToken = null
+            processQueuedSamples()
+        } catch (e: Exception) {
+            Log.e(TAG, "processSampleRunnable() threw an unexpected exception: ${e.message}", e)
+        }
+    }
+
+    protected fun finalize() { // destroys an instance of RetryBackend
+        Log.d(TAG, "finalize")
+        scheduleSampleHandler.removeCallbacksAndMessages(null) // removes all callbacks and messages
+    }
```

Comment:
https://stackoverflow.com/questions/48623441/destructor-in-kotlin-programming-language
I found this regarding the finalize method in java/kotlin...
Any thoughts/experiences with this?

Reply:
Problem @sabe25 is that this is not being used inside one small `using` statement but rather for a longer kind and when it gets collected we need to make sure these callbacks get unregistered too. I think this usage is safe here